neutron_plugin/router.py

```diff
@@ -87,69 +95,76 @@ def create(neutron_client, args, **kwargs):
-                result[k] = deepcopy(v)
-        return result
-
-    # Find out if the update script is being called
-    # from a relationship or a node operation.
-    if ctx.type == RELATIONSHIP_INSTANCE:
-        if ROUTER_OPENSTACK_TYPE in get_openstack_type(ctx.source):
-            subject = ctx.source
-        elif ROUTER_OPENSTACK_TYPE in get_openstack_type(ctx.target):
-            subject = ctx.target
+def update_routes(neutron_client, args, **kwargs):
+    routes = args.get(ROUTES_OPENSTACK_TYPE)
+    if not routes:
+        raise NonRecoverableError('routes param is required and must be '
+                                  'provided when creating static routes !!')
+    # Force to pass only the "routes" provided by the node properties
+    routes_args = {'routes': routes}
+
+    # This will update the router and add new static routes based on the
+    # routes param provided by the "cloudify.openstack.nodes.Routes"
+    r = _update_router_routes(neutron_client, routes_args, **kwargs)
+    router = r.get(ROUTER_OPENSTACK_TYPE)
+    if r and router:
+        # If the current context type is a relationship then update the
+        # source instance "runtime_properties" otherwise just update the
+        # current instance "runtime_properties"
+        if ctx.type == RELATIONSHIP_INSTANCE:
+            ctx.source.instance.\
+                runtime_properties[ROUTES_OPENSTACK_TYPE] = routes
         else:
-            raise NonRecoverableError(
-                'Neither target nor source is {0}'.format(
-                    ROUTER_OPENSTACK_TYPE))
+            ctx.instance.runtime_properties[ROUTES_OPENSTACK_TYPE] = routes
     else:
-        subject = ctx
-
-    try:
-        router = neutron_client.show_router(get_openstack_id(subject))
-    except NeutronClientException as e:
-        raise NonRecoverableError('Error: {0}'.format(str(e)))
-    if not isinstance(router, dict) or \
-            'router' not in router.keys() or \
-            'id' not in router['router'].keys():
         raise NonRecoverableError(
-            'API returned unexpected structure.: {0}'.format(router))
+            'Failed while trying to retrieve router instance')
 
-    router_id = router['router'].pop('id')
 
-    new_router = {'router': {}}
-    for key, value in args.items():
-        new_router['router'][key] = value
+@operation
+@with_neutron_client
+def add_routes(neutron_client, args, **kwargs):
+
+    # Since routes is part of router and not single API resource for routes
+    # "router" resource is used
+    router = use_external_resource(ctx, neutron_client, ROUTER_OPENSTACK_TYPE)
+    if router:
+        # Update routes as part of runtime properties
+        ctx.instance.runtime_properties[ROUTES_OPENSTACK_TYPE]\
+            = router[ROUTES_OPENSTACK_TYPE]
+        # Update type to match it as routes types
+        ctx.instance.runtime_properties[OPENSTACK_TYPE_PROPERTY]\
+            = ROUTES_OPENSTACK_TYPE
+        return
```

Comment:
I am wondering if in the case of an "existing route" we should verify that it actually exists? As in the route contents are true? What if a user says "use route B", but what they think is "route B is actually Route C"?

Reply:
This is just verify if the router exists, cannot check if the routes exists or not